User input

dboone opened the pull request:
Added `parse_input(...)` function which converts a string containing the file and rank positions of a move to a sequence of board indices.

Closes #6.

---

Todo:
- ~~Replace the tuples of `usize` with the `BoardPosition` struct.~~
- ~~Add tests to cover more than 2 move positions.~~

dboone committed:
Added function to get user input

And supporting functions. @drbassett and I have gone over this code
several times to make my initial implementation more rusty. In the
future, I need to commit my work in progress so we can see just how far
we have come!

dboone committed:
Updated Rust version in Travis CI configuration

We require Rust v1.1.0 for the `std::str::split_whitespace()` function.

dboone committed:
Made `BoardPosition` public

Also added the Debug attribute.

dboone committed:
Updated input functions to return `BoardPosition`

Also updated tests to use `BoardPosition` and added a three position
test.

drbassett commented:
Commit c9ea4bcecc96b10f696a28a96c4f2fc58d50bb29 failed to build in Travis because it couldn't find a Rust 1.1 function, but the pull request says that all checks have passed. Interesting. Do you think Travis just took a bit to pick up the Rust version change?

Edit:

Nevermind, I just realized you updated the Rust version in the commit _after_ the one that failed. Such is the danger of working with tools that have a differing version from the build.

drbassett commented:
The tests in [this file](https://github.com/dboone/rusty-checkers/blob/53d18341c0d7e600ea9c132e133b320a706ed392/src/checkers/input.rs) do not test any of the error conditions. A complete set of tests includes error conditions; that an error occurs when it is supposed to, and the right type of error, since we use `enum`s to identify different types of errors. I don't feel comfortable merging this in until we know the errors work as intended. Otherwise, the user may get the wrong error message, or none at all!

dboone commented:
Can we use `corrosion` to provide parameterized tests of the error conditions? My old tests covered the error conditions so I'm confident in them, but we do need to test them.

In GitLab if the merge request isn't assigned, it isn't ready to be merged.

drbassett commented:
We don't need to add any more utilities to do testing - we have everything we need to test the error conditions. Errors are just values, so `assert_eq!` can be used on them, assuming the error types have all implemented the equality trait (whatever it is called). Is there any reason we can't use `assert_eq!`?

dboone commented:
Where are you coming from with "We don't need to add any more utilities"?

I think you answered my question though. We can _continue_ using `corrosion` to test the error conditions.

drbassett commented:
Oh, you threw me off when you mentioned `corrosion` library in general - I thought you were implying that we needed to add something to `corrosion` to test for errors. If you had asked if we could use `ptest` specifically to test the error conditions, I would have been less confused.

dboone committed:
Added `InputError` tests

Tests all current error states:
* `TooFewTokens`
* `MissingRank`
* `MissingFile`
* `ZeroRank`
* `InvalidCharacter`

dboone committed:
Added empty string test

dboone committed:
Rearrange test functions

Calls to `ptest!(...)` are now immediately after the function they call.

dboone committed:
Updated `InputError` and `TokenError` attributes

Combined the `derive` attributes into one line.

dboone committed:
Remove unused variable `readFile`

This variable was left over from the original iteration of the
`parse_file_rank()` function. Not sure why it didn't have a `snake_case`
name!

dboone commented:
Removed the unused mutable variable `readFile`. There are no more warnings originating from the `input.rs` file.
